クラス(JavaScript)

jun-kondo opened the pull request:
メモアプリコマンドライン版

jun-kondo committed:
ライブラリの追加

メモアプリ作成に必要なライブラリをインストール

jun-kondo committed:
メモアプリ作成

まとめてコミットになってしまった。。。

jun-kondo committed:
[modify]クラス名変更 MemoService→MemoCrud

memoアプリのビジネスロジック部分のクラス

jun-kondo committed:
[modify]メモ作成成功時のメッセージを追加

保存時の成功と失敗の処理のハンドリングを追加

jun-kondo committed:
[modify]変数名の変更 memoService→memoCrud

前々回のコミットの変更漏れ

yuuu commented on 04.class/db_manager.js:
```diff
@@ -0,0 +1,104 @@
+import sqlite3 from "sqlite3";
+
+export class DbManager {
```
クラス名が少し具体過ぎますね。このクラスが持っている `sqlite3.Database` を1段階抽象化した名前にしましょう。このクラスでは `memos` テーブルを操作対象としている点もヒントになると思っています 👍 

ちなみに、 `er` や `tor` で終わるクラス名は「処理」に着目しているケースが多いので、出てきたら注意が必要です。処理ではなくそのオブジェクトを表す名前になっているか再確認するようにしましょう。
ただし、Railsにおける `Controller` は例外です。これはMVCから取ってきている名前なので 🙏

jun-kondo replied:
動詞+er/orはよろしくなかったです💦

> このクラスが持っている sqlite3.Database を1段階抽象化した名前にしましょう。このクラスでは memos テーブルを操作対象としている点もヒントになると思っています 👍

データベースより上位の概念でストレージかな？と思い変更してみました。いかがでしょうか👀？
8e54dfa12a7e97f11c2e4d7e73acf761de3e3f84

yuuu commented on 04.class/db_manager.js:
```diff
@@ -0,0 +1,104 @@
+
+  getAllTitles() {
+    return new Promise((resolve, reject) => {
+      this.#db.all("SELECT title FROM memos", (err, rows) => {
```
SQLの仕様として、orderがないクエリの結果は順番が保証されないので、並び順が毎回変化する懸念があります。明示的にorderするようにしましょう。

yuuu replied:
SELECTする対象はシチュエーションに応じて絞り込んだ方が良いでしょうか？全カラムを取得するコストはそこまで大きくないので、毎回全カラム取得して、アプリ側で必要なカラムのみ参照する、という方式だと問題がありますか？
SQL実行時の僅かなコストと引き換えに、コードはかなりスッキリすると思っています。

jun-kondo replied:
> 明示的にorderするようにしましょう。

id順にソートして表示するように`ORDER BY id`を追加しました🙏

> SELECTする対象はシチュエーションに応じて絞り込んだ方が良いでしょうか？

全レコードを取得するメソッドを取得カラムに応じて2つ使い分けていましたが、1つに統一して呼び出し元で参照を絞るようにしました👀
bc3cb91fb4e78b4863edee085ef5df95c7900498

yuuu commented on 04.class/memo_crud.js:
```diff
@@ -0,0 +1,101 @@
+      console.log("There are no registered memos.");
+      return;
+    }
+    const prompt = new Select({
```
メモを選択する処理が重複しているので、共通化ができないか検討をお願いします。

jun-kondo replied:
メモ選択処理を`selectMemo()`に切り出しました。
59e2ba9d51105aa6914207e0bf03559cde2a338c

yuuu commented on 04.class/memo_crud.js:
```diff
@@ -0,0 +1,101 @@
+      const result = await this.#dbManager.insertMemo(memo);
+      console.log(result);
+    } catch (error) {
+      console.error("Error registering new memo: ", error);
```
catchの処理はどれも基本的にエラーメッセージを出力しているだけなので、このクラスではメッセージ付きの例外を `throw` するだけにして、呼び元にてcatch→出力するようにすると、catchの処理を共通化できそうですね。

jun-kondo replied:
1. console.errorを呼び出し元の`run()`のcatch節に限定しました。
2. 呼び出し先のメソッドのcatch節でDBの操作のエラーを捕捉し、throw new Errorするようにしました。
上記のように修正してみましたが、如何でしょうか？

疑問点がありまして、現状エラーが起きると2回catch節を通るのですが、書き方的に1つのcatchで捌くべきでしょうか？
catchでエラーを投げているのですが、書き方的にこれは大丈夫かわかりませんでした。。。
70118a1f5d9688ce2d0a542bfd174c7dd7be5ec1

yuuu commented on 04.class/memo_crud.js:
```diff
@@ -0,0 +1,101 @@
+
+const { Select } = pkg;
+
+export class MemoCrud {
```
このクラスの役割が少し薄いかもと感じています。`MemoApp` と `DbManager` に責務を分散することでこのクラスをなくせないか検討をお願いします。

jun-kondo replied:
こちら対応しました。
もともと`MemoApp`にあったメソッドを切り出して作ったクラスだったので、言われてみるとモジュールっぽい役割になっているなと思いました。。。
そんな経緯もあり`DbManager`に移すメソッドがしっくりこず、全部`MemoApp`に移動してしまったのですがいかがでしょうか?
2af3337b22123db85837f237e61f1c647e5987f9

yuuu commented on 04.class/command_parser.js:
```diff
@@ -0,0 +1,18 @@
+import { Command } from "commander";
+
+export class CommandParser {
```
このクラス名も `CommandOption` といった名前の方が良いですね。
そうするとメソッド名も `parse` といった自然な動詞を選択できるようになります。

jun-kondo replied:
こちら修正しました!! 
cf305c4

jun-kondo committed:
[modify]命名をCommandOptionクラスに変更

パースするインスタンスメソッドをparseメソッドに命名変更

jun-kondo committed:
[modify]メソッド getAllMemosに命名変更

全メモを取得するメソッド名を変更し、全カラムのデータを取得するようにした。
役割が重複しているgetAllMemoIdsAndTitlesメソッドを削除し統合。

jun-kondo committed:
[modify]クラス名変更 MemoStorage

DbManagerから変更

jun-kondo committed:
[add]メソッド追加 selectMemo()

選択肢プロンプトを出力する処理を切り出した

jun-kondo committed:
[modify]catch節の処理 throw new Errorに変更

DB操作処理でのエラーをcatch節で捕捉し、新しくErrorオブジェクトを投げるように変更。
上位処理のcatch節でエラーメッセージを出力するようにする

jun-kondo committed:
[delete]memo_crudクラス

クラスの役割をmemo_appに移動、memo_crudクラスを削除

jun-kondo committed:
[modify]インスタンス変数名 memoStorage

変数に代入するオブジェクトのクラス名がMemoStorageなので整合性が取れるよう変更

jun-kondo committed:
[modify]console.logの使用箇所を限定

各メソッドで文字列を出力するようにしていたのを呼び出し元で行うように変更
メソッドは必ず戻り値をreturnするようにする

jun-kondo committed:
[delete]インスタンス変数

消し忘れのインスタンス変数を削除

jun-kondo committed:
[add]空行追加

jun-kondo committed:
[add]例外追加 オプションが2個以上入力された場合

`-rl`のような入力があった場合例外をスロー

jun-kondo committed:
[Modify]表示順の変更

jun-kondo commented:
お疲れ様です。レビューありがとうございます🙌
修正して返信いたしました。

ご指摘箇所以外にも、
出力箇所(console.log)を呼び出し元で限定するようにして、呼び出し先は戻り値をreturnするように修正 702b65870bdd99852e30eb71eec668cfb0e098ca
オプションを2つ以上入力しても動作してしまうので、その場合例外を投げるように処理を追加 997cfd3ac701a8724389e6276d8cbc1abb3f3e84
が気になったので変更してみました!!
恐れ入りますがご確認お願いいたします!!

jun-kondo committed:
[Modify]エラー処理 SQLITEのエラーのみ捕捉

データベース処理時のエラーのみcatchしてそれ以外は例外を投げる

jun-kondo committed:
[Delete]エラー処理 オプションが複数渡されたとき

エラーではなく、複数オプションが使えない旨のメッセージを出力するよう変更

jun-kondo committed:
[Modify]メモが1つもなかったときの処理

戻り値がundefinedになっていたので、メモが無い旨のメッセージを返すように変更

jun-kondo committed:
[Modify]Selectプロンプトからメモのidを返却するように変更

戻り値がメモのタイトルになっていたため、タイトルが同じメモの区別ができなくなっていた

jun-kondo commented:
こちら再度修正しました。

> オプションを2つ以上入力しても動作してしまうので、その場合例外を投げるように処理を追加 [997cfd3](https://github.com/jun-kondo/js-practices/commit/997cfd3ac701a8724389e6276d8cbc1abb3f3e84) が気になったので変更してみました!!

上記の件ですが例外を発生させずに条件分岐で2個以上のオプションを受け取っていた場合は早期リターンしてそれ以上処理を進めないように変更しました。
[f3c6195](https://github.com/jun-kondo/js-practices/pull/4/commits/f3c619520eab8579ca9ed663da230c1130ee1f0e)

yuuu commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,114 @@
+      if (e && e.code === "SQLITE_ERROR") {
+        console.error(e.message);
+      } else {
+        throw e;
```
https://github.com/jun-kondo/js-practices/pull/4#discussion_r1673772155 で言及されている箇所はおそらくここですよね？「想定されたエラーであれば内容を出力して、より重大なエラーであればエスカレーションする」という意図だと思うので、こうした実装自体には問題はないかなと思います。

一方でここの分岐は `code` で行うのではなく条件付きcatchブロックを使った方が、確実ですね。 `code` という属性を持っている例外がthrowされてくるとは限らないので。
https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Statements/try...catch#%E6%9D%A1%E4%BB%B6%E4%BB%98%E3%81%8D_catch_%E3%83%96%E3%83%AD%E3%83%83%E3%82%AF

jun-kondo replied:
> 「想定されたエラーであれば内容を出力して、より重大なエラーであればエスカレーションする」という意図だと思うので、こうした実装自体には問題はないかなと思います。

ありがとうございます。自分でも考えて見ましたがcatch節に想定外のエラーが来たらthrowしないと見逃すことになるので納得しました!!

> 一方でここの分岐は code で行うのではなく条件付きcatchブロックを使った方が、確実ですね。 code という属性を持っている例外がthrowされてくるとは限らないので。

頂いたURL確認しました。
`e instanceof Error`でErrorのインスタンスで判別して分岐するであっているでしょうか？
捕捉したい想定エラーがデータベース操作時のエラー(Errorオブジェクト)になります。
単純に`e instanceof Error`で分岐すると、継承先のエラー(RangeErrorやTypeError)も`true`になってしまうので、`message`のプロパティの内容を照合する条件を追加しました。
Errorオブジェクトの`name`や`message`のプロパティであればアクセス出来ると想定できるのでこのようにしましたが如何でしょうか?👀

小さくて見辛いかもですが、想定しているエラーオブジェクトをデバッグで調べてみました。
<img width="1305" alt="image" src="https://github.com/user-attachments/assets/d1a619f2-ab7c-4588-bf17-cc608103fff8">

yuuu replied:
なるほど、 [node-sqlite3](https://github.com/search?q=repo%3ATryGhost%2Fnode-sqlite3%20SQLITE_ERROR&type=code) のコードも見てみましたが、確かにthrowされるのは単なるErrorオブジェクトのようですね 💦 確認ありがとうございました。この方法でOKです。

jun-kondo committed:
[Modify]条件付きcatch節

(e instanceof Error && e.message.includes("SQLITE_ERROR"))でDB操作時のエラーを捕捉する
